### ISSUE TITLE
neural_renderer is not differentiable？

### ISSUE THREAD

**Issue by xhuan8** [akashsengupta1997/STRAPS-3DHumanShapePose#16]
I try to use silhouette to supervise the training, but the result did not converge
any suggestion?

**Comment by akashsengupta1997** [comment]
IIRC, the NMRRenderer class I used for training data generation is not differentiable (but I can't remember why now), you will need to define a new renderer e.g.

`nmr_for_silhouette_loss = nr.Renderer(camera_mode='projection',
                                          K=cam_K,
                                          R=cam_R,
                                          image_size=img_wh,
                                          orig_size=img_wh).to(device)`

and then get predicted silhouettes in the training loop:
` pred_silhouettes = nmr_for_silhouette_loss(vertices=pred_vertices, faces=faces,
                                                           t=pred_cam_ts.unsqueeze(1),
                                                           mode='silhouettes')`

Actually the `mode=silhouettes` argument might be the key to making it differentiable.

However, I would suggest not using NMR at all as a differentiable renderer, and looking at pytorch3d (https://github.com/facebookresearch/pytorch3d) instead.

**Comment by xhuan8** [comment]
thanks for the quick reply, the code you provided also not working.
I'll try PyTroch3D, but the parameters of PyTroch3D are not the same, take some time to figure it out

**Comment by zhihaolee** [comment]
Hi @akashsengupta1997 ， may I ask why you suggest not using NMR as a differentiable renderer, but pytorch3d instead? Thank you!

**Comment by akashsengupta1997** [comment]
Hi @zhihaolee, I believe that pytorch3d is a newer library with better maintenance, documentation and support/bug-fixes (the commonly-used port of NMR to pytorch is no longer updated). Furthermore, pytorch3d has extensive tutorials and well-commented code that is fairly easy to understand. And perhaps most importantly, pytorch3d implements more recent differentiable rendering frameworks (e.g. SoftRas: https://github.com/ShichenLiu/SoftRas) which should perform better than NMR in practice.

**Comment by zhihaolee** [comment]
Thanks for your reply, @akashsengupta1997 .

I agree with you, and I found that there may be problems in the gradient calculation of NMR, as mentioned at https://github.com/daniilidis-group/neural_renderer/issues/18 .

**Comment by akashsengupta1997** [comment]
Yes that is a very good point also! I believe that the NMR gradients are approximated using "hand-crafted" functions (as detailed in the SoftRas paper), so I can imagine there may be a higher chance of bugs in implementations.
